spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSessionCatalog.java

```diff
@@ -134,28 +135,34 @@ public Identifier[] listTables(String[] namespace) throws NoSuchNamespaceExcepti
 
   @Override
   public Table loadTable(Identifier ident) throws NoSuchTableException {
-    try {
-      return icebergCatalog.loadTable(ident);
-    } catch (NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident);
-    }
+    return loadTableInternal(ident, null, null);
   }
 
   @Override
   public Table loadTable(Identifier ident, String version) throws NoSuchTableException {
-    try {
-      return icebergCatalog.loadTable(ident, version);
-    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident, version);
-    }
+    return loadTableInternal(ident, version, null);
   }
 
   @Override
   public Table loadTable(Identifier ident, long timestamp) throws NoSuchTableException {
+    return loadTableInternal(ident, null, timestamp);
+  }
+
+  private Table loadTableInternal(Identifier ident, String version, Long timestamp)
+      throws NoSuchTableException {
     try {
-      return icebergCatalog.loadTable(ident, timestamp);
+      return loadTable(icebergCatalog, ident, version, timestamp);
     } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident, timestamp);
+      return loadTable(getSessionCatalog(), ident, version, timestamp);
+    } catch (org.apache.iceberg.exceptions.NotFoundException e) {
+      if (loadCatalogTableWhenMetadataNotFoundEnabled()) {
```

Comment:
It feels a bit hacky to me. I wonder whether that unregister table procedure may be a better option.
Let me think.

Reply:
We discussed possibly just overriding tableExists so that if we have a "unable to read file exception" from loadTable we can still say the table exists as an Iceberg table. Further catalog attempts to use the table would fail, but delete would succeed. If we want we could also gate that behind session param.

Reply:
I think we are trying to fix this problem at a wrong level. I feel Spark logic should be as follows:

- Load Iceberg table if possible. If not possible, it is OK.
- Call icebergCatalog.drop(ident, false) to drop the pointer no matter whether the first step was successful.
- If Iceberg table is not null and metadata file still exists, call the cleanup action.

All we would need to do is to adapt our Iceberg catalog implementations to drop pointers even if the metadata file does not exist. That's where we can check if the catalog entry seems like an Iceberg table.

Reply:
I don't think we need any Spark session configs for that. It is up to Iceberg catalogs to verify it is a valid Iceberg table. We can't make the assumption on how the catalog handles Iceberg metadata.

Reply:
The Spark execution plan requires calling "tableExists" before an invocation of "dropTable" so we can't fix it purely within our implementation unless we also adjust "tableExists" That's why I had suggested fixing "tableExists" so that it doesn't fail just because "loadTable" also failed, but would return true as along as the exception had to do with something Iceberg related like attempting to load metadata.

Reply:
Thanks @RussellSpitzer . 

hi @aokolnychyi. I agree with Russell. Can I make the change suggested by Russell? Please advise. Thanks

Reply:
hi @aokolnychyi @RussellSpitzer I tried to move this logic to tableExists method, but it looks like that also wont work.  Spark calls loadTable when it tries to convert Unresolvedrelation to resolved table in 

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L1151

```
      case u @ UnresolvedTableOrView(identifier, cmd, allowTempView) =>
        lookupTableOrView(identifier).map {
          case v: ResolvedView if v.isTemp && !allowTempView =>
            throw QueryCompilationErrors.expectTableOrPermanentViewNotTempViewError(
              identifier.quoted, cmd, u)
          case other => other
        }.getOrElse(u)

    private def lookupTableOrView(identifier: Seq[String]): Option[LogicalPlan] = {
      lookupTempView(identifier).map { _ =>
        ResolvedView(identifier.asIdentifier, isTemp = true)
      }.orElse {
        expandIdentifier(identifier) match {
          case CatalogAndIdentifier(catalog, ident) =>
            **CatalogV2Util.loadTable(catalog, ident)**.map {
              case v1Table: V1Table if CatalogV2Util.isSessionCatalog(catalog) &&
                v1Table.v1Table.tableType == CatalogTableType.VIEW =>
                ResolvedView(ident, isTemp = false)
              case table =>
                ResolvedTable.create(catalog.asTableCatalog, ident, table)
            }
          case _ => None
        }
      }
    }
```

Could you please advise what else I can try here. TIA

Reply:
Okay, I see `DropTableExec` implementation now. I agree about differentiating between table exists vs being able to load table metadata. We can change `tableExists` in Iceberg catalogs to verify if the metastore pointer exists rather that trying to load the table and then overload `tableExists` inside Spark catalogs.

I feel we can still implement what I suggested [above](https://github.com/apache/iceberg/pull/7228/files#r1157829683) for the actual cleanup.

@dpaani, that's unfortunate. One idea to overcome that issue is to throw a new exception type like `CorruptedMetadataException` in our Iceberg catalogs whenever the metadata file is missing. We can then catch this inside our Spark catalogs and return a special Spark table.

Let me think more tomorrow. Any better ideas, @RussellSpitzer?